Minimal BIP9 implementation

sipa opened the pull request:
This is an alternative to #6816 and #7566, implementing BIP9 with minimal code changes, including mining logic, softfork warning logic, and some unit tests.

It adds support for a begin time before version bits are counted, which is not in BIP9 currently, though it is compatible.

The last commit demonstrates how to add a simple softfork.

sipa committed:
BIP9 Implementation

Inspired by former implementations by Eric Lombrozo and Rusty Russell, and
based on code by Jorge Timon.

sipa committed:
Versionbits tests

btcdrak commented:
Could you add status to the `getblockchaininfo` RPC?

dcousens commented:
@sipa any reason not to just extend #7566?

btcdrak commented:
Concept ACK, I prefer this BIP9 implementation.

sipa commented:
@dcousens I wrote most of this before I saw #7566, which would need some
refactoring anyway to support the warning logic. I did use the params
structure from there.

jonasschnelli commented:
There are some `versionbits_test.cpp` formatting/syntax issues (https://travis-ci.org/bitcoin/bitcoin/jobs/110833452#L1736)

sipa commented:
Fixed tests.

jtimon commented on src/main.cpp:
```diff
@@ -2080,6 +2081,47 @@ void PartitionCheck(bool (*initialDownloadCheck)(), CCriticalSection& cs, const
+        }
+    }
+
+    if (nVersion == 0) {
```
Why not simply start using the versionbit prefix without voting for any concrete bit instead?

sipa replied:
> Why not simply start using the versionbit prefix without voting for any
concrete bit instead?

That would mean that versionbits on itself is a change, and lack of
adoption of versionbits by miners would then cause the warning logic to
trigger on new nodes.

Using this logc here, versionbits gets rolled out simultaneously with the
first fork that uses it.

jtimon replied:
I mean, if we are going to wait to deploy the first SF to merge it, I think it makes even more sense that we never return Consensus::VERSIONBITS_LAST_OLD_BLOCK_VERSION here.

>> That would mean that versionbits on itself is a change, and lack of
adoption of versionbits by miners would then cause the warning logic to
trigger on new nodes.

Just like changing the version to 8 without doing anything else would be. But it doesn't affect functionality.
The new nodes would only get warnings if the miners, apart from adopting this change were adopting or implementing other change to vote on concrete bits (if they only set nVersion to Consensus::VERSIONBITS_TOP_BITS), no warnings should be raised (unless an unknown SF has been deployed using versionbits, in which case is good that they get the warnings).

sipa replied:
Assume the start date for the first bip9 softfork is april 1st, but we
release code with bip9 and that softfork on march 1st.

That means that it's likely that during march no new-style nVersion will be
mined. That also means that anyone who does run upgraded code will get
warnings, because the nVersion they expect (0x20000000) differs from the
one they see (4).

jtimon replied:
Ok, I get it, it's the old nodes seeing the "empty versionbits" version which would get (unnecessarily?) be told to upgrade and it doesn't matter whether we introduce a SF with BIP9 or not because the start date will be after the miners start using the new code to produce their new versions.

But I think I still disagree, though.
If in march 15th 95% of the miners are ready to deploy a softfork scheduled for april 1st, I believe old nodes should actually receive the upgrade warning already, even if the message is not accurate.

jtimon commented on src/main.cpp:
```diff
@@ -2451,20 +2498,14 @@ void static UpdateTip(CBlockIndex *pindexNew) {
     static bool fWarned = false;
     if (!IsInitialBlockDownload() && !fWarned)
```
I know this is not changed in this PR, but why don't we want these warnings on initial block download?
Would it be a problem to make these checks in ConnectBlock() rather than UpdateTip() ?

jtimon commented on src/main.cpp:
```diff
@@ -2451,20 +2498,14 @@ void static UpdateTip(CBlockIndex *pindexNew) {
             // strMiscWarning is read by GetWarnings(), called by Qt and the JSON-RPC code to warn the user:
-            strMiscWarning = _("Warning: This version is obsolete; upgrade required!");
+            if (warningcheck.IsActiveFor(pindex, chainParams.GetConsensus())) {
+                strMiscWarning = _("Warning: unknown softfork has activated");
```
Is it really worth it to notify this twice (once when is locked in and 2016 blocks later when it's activated)?

sipa replied:
Well, one is before you're forked off and the other is after. The warning
texts here could certainly be more descriptive.

jtimon replied:
But unless there's a reorg, nothing is going to stop the locked in deployment from being activated.
I guess more messages don't hurt, even if we need to repeat the check (states are cached anyway).

jtimon commented on src/consensus/versionbits.h:
```diff
@@ -0,0 +1,96 @@
+    // A map that gives the state for blocks whose height is a multiple of Period().
+    // The map is indexed by the block's parent, however, so all keys in the map
+    // will either be NULL or a block with (height + 1) % Period() == 0.
+    std::map<const CBlockIndex*, State> cache;
```
This is the main difference between your implementation and mine. You have a separate Cache for each deployment while I have a single cache with containing the states of all the deployments. While mine should be more efficient (at least storage-wise), your code is more reusable (as shown in the way you implement the warnings checks).
Another goal for caching the states together is to simplify things in case libconsensus exposes (C API) the cache in the future to support concurrent calls.

jtimon commented on src/consensus/versionbits.h:
```diff
@@ -0,0 +1,96 @@
+
+    enum State {
+        UNDEFINED,
+        DEFINED,
```
Bike-shedding: I'm doing s/UNDEFINED/DEFINED s/DEFINED/STARTED isntead, which I think it's more clear:
If it's in the chainparams is defined, it just hasn't started yet. Undefined/unkown deployments are those which you don't have in your chainparams.

sipa replied:
Makes sense.

sipa replied:
See https://github.com/bitcoin/bips/pull/344.

jtimon commented on src/consensus/versionbits.cpp:
```diff
@@ -0,0 +1,95 @@
+                if (pindexPrev->GetMedianTimePast() >= nTimeTimeout) {
+                    stateNext = FAILED;
+                } else if (pindexPrev->GetMedianTimePast() >= nTimeDefined) {
+                    stateNext = DEFINED;
```
Shouldn't this check that the same bit is not currently being used by other deployment defined in chainparams for the same bit (and already started or locked in) before advancing to the next state?

jtimon commented on src/consensus/versionbits.cpp:
```diff
@@ -0,0 +1,95 @@
+
+        switch (state) {
+            case UNDEFINED: {
+                if (pindexPrev->GetMedianTimePast() >= nTimeTimeout) {
```
This check is redundant with the similar one in the DEFINED state (assuming the check is done before counting upgrade confirmations). You could also remove the break, so that a deployment changing from UNDEFINED to DEFINED has the chance of moving to FAILED or LOCKED_IN in the same round.

sipa replied:
I think the semantics are simpler without the possibility of going from
undefined to lockedin in one transition.

jtimon replied:
fair enough, we're talking about saving 3-4 lines to libconsensus or leave them as self-documentation. If I forget to copy you on this, please, remind me.

jtimon commented on src/main.cpp:
```diff
@@ -2080,6 +2081,47 @@ void PartitionCheck(bool (*initialDownloadCheck)(), CCriticalSection& cs, const
     }
 }
 
+Consensus::VersionBitsConditionChecker checkCSV(Consensus::DEPLOYMENT_CSV);
```
These could be static and be in consensus/versionbits.cpp instead of main.cpp by moving ComputeBlockVersion() to versionbits as well (of course, without cs_main).

WarningBitsConditionChecker could be encapsulated behind a function there too. In a very far future in which Bitcoin Core only uses libconsensus through its C API, that ```GetVersionBitsWarnings(const CBlockIndex* pindexPrev, const Consensus::Params&)``` function seems much closer from being "exposable" in libconsensus than class ```Consensus::AbstractThresholdConditionChecker```.

I know that ComputeBlockVersion() and GetVersionBitsWarnings() are not strictly consensus code, but they heavily depend on consensus code, and I think it would be a wasted opportunity not to have this functionality as encapsulated as possible from birth.

sipa replied:
The reason I didn't put them in versionbits is exactly because the lock is
in main. Post c++11 we can move locks to libconsensus, and the versionbits
cache can have its own lock.

jtimon replied:
I don't think we want locks in libconsensus, not even post c++11. This reminds me of a previous conversation we had about concurrency and the cache. It seems a renounced to supporting concurrent calls without locks in libconsensus for the wrong reason. When you previously said " no need to have locks in the internal cache, the caller can get cs_main", it seems you meant something different than what I understood. I will go back to having a C API interface for the cache.

In this PR, can't the caller of GetFlags (), ComputeBlockVersion() and GetVersionBitsWarnings() take the lock (assuming you move this code behind those 3 functions to encapsulate the static vars)?

sipa replied:
I'll try another implementation that does that. Separating the cache and the logic probably makes sense.

jtimon commented on src/main.cpp:
```diff
@@ -2176,6 +2218,11 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
         flags |= SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY;
     }
 
+    // Start enforcing CHECKSEQUENCEVERIFY using versionbits logic.
```
please, please, please libconsensus refactor nit:

Could you replace these lines with
``` 
flags |= Consensus::GetFlags(pindex, chainparams.GetConsensus());
```
?
(is it ```pindex->pprev``` or just ```pindex``` here?)

That implies, again, that checkCSV (and thus softforkcheck) would need to move to consensus/versionbits.cpp.

sdaftuar replied:
I don't have a strong opinion here, but I don't know that it really makes sense to call a generic `GetFlags()` function, as I believe we will need to have a separate one of those to deploy BIP68 (which doesn't use a script flags variable, but uses a locktimeflags variable instead)

jtimon replied:
I meant by unifying the consensus flags (currently only script and "locktime"), see https://github.com/jtimon/bitcoin/commit/b77f1f21eda2329d99d7f2c3242be228db58b131 .

jtimon commented on src/consensus/versionbits.cpp:
```diff
@@ -0,0 +1,95 @@
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "versionbits.h"
+
```
missing line
```
#include "chain.h" // TODO Decouple libconsensus from CBlockIndex.
```

sipa commented:
Updated to correspond to latest BIP9 (see https://github.com/bitcoin/bips/pull/344, https://github.com/bitcoin/bips/pull/345, https://github.com/bitcoin/bips/pull/346), and fixed the tests.

I also have an alternate version that factors out the cache, see https://github.com/sipa/bitcoin/commit/218d2fdcbfb207ed75f4f5a6fc7a1b5de6d791b8).

sipa committed:
Softfork status report in RPC

sipa commented:
1) Moved the versionbits logic out of consensus, and make it not depend on it is
2) Added (very basic) reporting of BIP9 deployment status through RPC
3) Switched the window size from 2016 to 144 for regtest

Should be ready for review.

jtimon commented:
utACK 7662816

btcdrak commented on src/main.cpp:
```diff
@@ -2448,25 +2498,17 @@ void static UpdateTip(CBlockIndex *pindexNew) {
+                strMiscWarning = _("Warning: unknown softfork has activated");
+                CAlert::Notify(strMiscWarning, true);
+            } else {
+                LogPrintf("Warning: An unknown softfork is about to activate");
```
missing a trailing `\n`

sipa commented:
Fixed a few bugs:
* typo in the warning logic which made it trigger all the time
* newline at the end of unknown softfork warning log message
* bip9_softforks was not being added to the getblockchaininfo output

paveljanik commented on src/consensus/params.h:
```diff
@@ -22,6 +43,14 @@ struct Params {
     uint256 BIP34Hash;
+    /**
+     * Minimum blocks including miner confirmation of the total of 2016 blocks in a retargetting period,
+     * (nPowTargetTimespan / nPowTargetSpacing) wich is also used for BIP9 deployments.
```
wich -> which

sipa replied:
Fixed.

sdaftuar commented on src/test/versionbits_tests.cpp:
```diff
@@ -0,0 +1,185 @@
+    VersionBitsTester& TestStarted() {
+        for (int i = 0; i < CHECKERS; i++) {
+            if ((insecure_rand() & ((1 << i) - 1)) == 0) {
+                BOOST_CHECK_MESSAGE(checker[i].GetStateFor(vpblock.empty() ? NULL : vpblock.back()) == THRESHOLD_STARTED, strprintf("Test %i for DEFINED", num));
```
should be `strprintf("Test %i for STARTED", num)`

sipa replied:
Fixed.

sdaftuar commented on src/test/versionbits_tests.cpp:
```diff
@@ -0,0 +1,185 @@
+BOOST_AUTO_TEST_CASE(versionbits_test)
+{
+    for (int i = 0; i < 64; i++) {
+        // UNDEFINED -> FAILED
```
Perhaps this comments and the ones below should be updated?  UNDEFINED is not a state.

sipa replied:
Fixed.

btcdrak commented:
Would it be possible to emit debug logs for the softfork state changes?

sipa commented:
@btcdrak That would require breaking some layering, as the consensus logic never sees the transitions, and the threshold logic is used for both consensus and warnings.

btcdrak commented:
@sipa no problem, I drop the idea.

btcdrak commented:
Tested ACK a7f15a6

Part of my tests involved porting the softforks activation from #7561 and converting the RPC tests to check both enforcement and RPC bip9 status. Work done in https://github.com/bitcoin/bitcoin/compare/master...btcdrak:vb_68_112_113_1

sdaftuar commented on src/main.cpp:
```diff
@@ -2080,6 +2081,50 @@ void PartitionCheck(bool (*initialDownloadCheck)(), CCriticalSection& cs, const
+
+    bool Condition(const CBlockIndex* pindex, const Consensus::Params& params) const
+    {
+        return (pindex->nVersion & ~ComputeBlockVersion(pindex->pprev, params)) != 0;
```
Won't this have potential false alerts, if let's say half the blocks in a period are signaling some unknown bit i, while the other half are signaling some other unknown bit j?

Anyway it occurred to me that it might be nice to record which bits are being set that are unknown and write that to the debug log when we warn a softfork is about to activate (as a useful debugging/investigation aid), but I can see that is a little tricky.

sipa replied:
That's quite doable actually: have a 'warning' tracker per bit.

sdaftuar replied:
Oh it also occurs to me that we probably shouldn't count a block towards the warning unless its top bits match the `001` versionbits prefix?  Otherwise we could see warnings if we deploy versionbits+some soft fork, but miners are slow to pick it up and continue using old version numbers.

Anyway I like the idea of a warning tracker per bit...

sdaftuar commented on src/main.cpp:
```diff
@@ -2080,6 +2081,50 @@ void PartitionCheck(bool (*initialDownloadCheck)(), CCriticalSection& cs, const
+// Protected by cs_main
+static VersionBitsCache versionbitscache;
+
+int32_t ComputeBlockVersion(const CBlockIndex* pindexPrev, const Consensus::Params& params)
```
I think it'd be nice if we had test coverage for this function.  I believe that as it stands, if this code were busted, then I'm not sure any tests would necessarily fail?

So far the best idea I have is that when we deploy a new soft fork, we write a unit test that exercises this function and verifies the right bit will be set at the right time (using a fake mainnet chain like the one you built in the versionbits unit test). 

I think we should also have regtest RPC tests which exercise proposed soft forks as well, but I can't think of any way to write such tests to adequately test mainnet parameters...

sipa replied:
Perhaps it would be useful to have a 'test' softfork that is disabled (end
time before genesis, for example) everywhere but regtest, and enables some
very easy to test condition (e.g. coinbase must have exactly one output)?

jtimon replied:
A testing deployment sounds like a good idea.

sdaftuar replied:
@sipa I tried implementing a unit test for `ComputeBlockVersion()` using a dummy deployment; not sure adding a dummy softfork for testing is something we want to do, but if you're interested the code for that is here:

https://github.com/sdaftuar/bitcoin/commit/b8ec634651fbc5a41afc47a768de4357983d3b4f

If you think this is worth including in some form, feel free to grab it, or I am happy to separately PR this unit test later.

sipa replied:
Cherry-picked.

morcos commented on src/main.cpp:
```diff
@@ -2080,6 +2081,50 @@ void PartitionCheck(bool (*initialDownloadCheck)(), CCriticalSection& cs, const
+
+    if (nVersion == 0) {
+        // Pre-versionbits
+        return VERSIONBITS_LAST_OLD_BLOCK_VERSION;
```
I'm not really sure the intention here.  This will also happen after version bits is rolled out but between any active soft fork deployments.  Is there a reason not to just always `| VERSIONBITS_TOP_BITS` and forget about version 4 blocks.

morcos commented:
@sipa see https://github.com/sipa/bitcoin/pull/58 for suggestions on how to rework the warning logic.
It separates it into two different things.  
 - Something similar to the old warnings which just runs in the steady state and lets you know when blocks are happening that aren't what you expect and triggers an alert if this is over 50%.  This could apply to unknown BIP 9 softforks or any other change to the use of nVersion.
 - Something actually detects activation of unknown BIP 9 soft forks.  This runs even during IBD.  I think this is important so old code doesn't miss the transient signaling of a new soft fork if it happened to not be online, or so that if someone accidentally downloads old code and syncs it, they can detect they are missing something.

gmaxwell commented:
I think we want tracking during IBD for another reason: I think we'll want to disable mining (return error on GBT) after an unknown BIP9 fork locks in unless explicitly overridden (ignorebip9fork=height:bit in config); otherwise there is a rule-reversion-risk, where long after a fork locks some non-trivial portion of hashpower happens to downgrade to older code that doesn't support the existing locked forks.

sdaftuar committed:
Test versionbits deployments

sdaftuar committed:
Add testing of ComputeBlockVersion

sdaftuar commented on src/test/versionbits_tests.cpp:
```diff
@@ -0,0 +1,185 @@
+                           .Mine(14333, TestTime(30003), 0).TestActive()
+                           .Mine(24000, TestTime(40000), 0).TestActive();
+    }
+}
```
I wrote up a couple additional tests to try sanity checking the mainnet deployments:

```
diff --git a/src/test/versionbits_tests.cpp b/src/test/versionbits_tests.cpp
index 9de8461..c384f92 100644
--- a/src/test/versionbits_tests.cpp
+++ b/src/test/versionbits_tests.cpp
@@ -6,6 +6,7 @@
 #include "random.h"
 #include "versionbits.h"
 #include "test/test_bitcoin.h"
+#include "chainparams.h"
 
 #include <boost/test/unit_test.hpp>
 
@@ -180,6 +181,28 @@ BOOST_AUTO_TEST_CASE(versionbits_test)
                            .Mine(14333, TestTime(30003), 0).TestActive()
                            .Mine(24000, TestTime(40000), 0).TestActive();
     }
+
+    // Sanity checks of version bit deployments
+    const Consensus::Params &mainnetParams = Params(CBaseChainParams::MAIN).GetConsensus();
+    for (int i=0; i<(int) Consensus::MAX_VERSION_BITS_DEPLOYMENTS; i++) {
+        uint32_t bitmask = VersionBitsMask(mainnetParams, (Consensus::DeploymentPos)i);
+        // Make sure that no deployment tries to set an invalid bit.
+        BOOST_CHECK(bitmask < (uint32_t)VERSIONBITS_TOP_BITS);
+
+        // Verify that the deployment windows of different deployment using the
+        // same bit are disjoint.
+        // This test may need modification at such time as a new deployment
+        // is proposed that reuses the bit of an activated soft fork, before the
+        // end time of that soft fork.  (Alternatively, the end time of that
+        // activated soft fork could be later changed to be earlier to avoid
+        // overlap.)
+        for (int j=i+1; j<(int) Consensus::MAX_VERSION_BITS_DEPLOYMENTS; j++) {
+            if (VersionBitsMask(mainnetParams, (Consensus::DeploymentPos)j) == bitmask) {
+                BOOST_CHECK(mainnetParams.vDeployments[j].nStartTime > mainnetParams.vDeployments[i].nTimeout ||
+                        mainnetParams.vDeployments[i].nStartTime > mainnetParams.vDeployments[j].nTimeout);
+            }
+        }
+    }
 }
 
 BOOST_AUTO_TEST_SUITE_END()
```

Perhaps worth including something like that here?

sipa replied:
Sure! Make a commit somewhere?

sdaftuar replied:
https://github.com/sdaftuar/bitcoin/commit/471d10583b197556ff7d9ae4c30a258a8f3061ff

sdaftuar commented on src/main.cpp:
```diff
@@ -2080,6 +2081,50 @@ void PartitionCheck(bool (*initialDownloadCheck)(), CCriticalSection& cs, const
 }
 
+// Protected by cs_main
+static VersionBitsCache versionbitscache;
```
I think we should probably clear out these caches in `UnloadBlockIndex()`.

sipa replied:
Done.

sipa commented:
@gmaxwell The code now will only do checks after IBD, but it will detect unknown softforks that activated in the past.

btcdrak commented:
@sipa Now that there is a "TESTDUMMY" softfork, can I suggest removing `Add CHECKSEQUENCEVERIFY softfork through BIP9` sipa@2b4bb57 and we add it through #7648 separately (which has been rebase to this branch now).

morcos commented on src/main.cpp:
```diff
@@ -2456,18 +2507,32 @@ void static UpdateTip(CBlockIndex *pindexNew) {
         {
             // strMiscWarning is read by GetWarnings(), called by Qt and the JSON-RPC code to warn the user:
-            strMiscWarning = _("Warning: This version is obsolete; upgrade required!");
+            strMiscWarning = _("Warning: unknown block versions being mined on the network!");
```
I'd prefer if this warning was a little bit stronger.  Perhaps you could change it to:
`"Warning: Unknown block versions being mined! It's possible unknown rules are in effect"`  or something to that effect?

sdaftuar commented on src/main.cpp:
```diff
@@ -2456,18 +2507,32 @@ void static UpdateTip(CBlockIndex *pindexNew) {
+                    CAlert::Notify(strMiscWarning, true);
+                    fWarned = true;
+                } else {
+                    LogPrintf("%s: unknown new rules are about to activate (versionbit %i)\n", bit);
```
Is this missing a `__func__`?

sdaftuar commented on src/main.cpp:
```diff
@@ -2456,18 +2507,32 @@ void static UpdateTip(CBlockIndex *pindexNew) {
+                }
+            }
+        }
+        int32_t nExpectedVersion = ComputeBlockVersion(pindex->pprev, chainParams.GetConsensus());
```
(Sorry if this is a repost, I think github ate my first attempt to comment)

I think we should move this line into the for loop below; otherwise we'd generate a warning after a known soft-fork activates.

sipa commented:
All comments addressed, I think.

sdaftuar committed:
RPC test for BIP9 warning logic

btcdrak commented:
Tested ACK 532cbb2

sdaftuar commented:
ACK

If anyone is interested I also tested the warning logic with an RPC test (https://github.com/sdaftuar/bitcoin/commit/4c2e8ba2282e0948aa8680ab5e8283e5f8ef12f5)

sipa commented:
Included @sdaftuar's warning logic RPC test.

jtimon commented on src/rpc/blockchain.cpp:
```diff
@@ -604,6 +604,20 @@ static UniValue SoftForkDesc(const std::string &name, int version, CBlockIndex*
     return rv;
 }
 
+static UniValue BIP9SoftForkDesc(const std::string& name, const Consensus::Params& consensusParams, Consensus::DeploymentPos id)
```
Shouldn't this loop throught all the deployments defined in Consensus::Params and be called from somewhere?
Style: why is not indented?

sipa replied:
It's called once for every deployment below, similar to how SoftForkMajorityDesc is. There just are no deployments just yet.

Making this function iterate through all deployments would require putting knowledge of the names of all deployments in the consensus params, and would make it inevitable that the test deployment is listed.

sipa replied:
Style" what do you mean by not indented?

jtimon replied:
Mhmm, I can only find one instance of the string "BIP9SoftForkDesc" in the changes.

re itarate with knowledge of the names, what's wrong with 
```
for (int i = 0; i < MAX_VERSION_BITS_DEPLOYMENTS; ++i)
``` ?

sipa replied:
You don't know what string name to give to the Object entry inside "bip9_softforks".

Yes you can't find any calls to it yet, because there are no BIP9 softforks defined. Future deployments using BIP9 are supposed to be added below, just like ISM softforks were supposed to be added below to "softforks".

jtimon replied:
Style, oh, never mind: https://github.com/bitcoin/bitcoin/blob/master/src/.clang-format#L27

I see, I didn't thought about the name.

jtimon commented:
re-utACK modulo last nit.

morcos commented:
Code review ACK

btcdrak commented:
Tested ACK 8c74ced

morcos commented:
@sipa I realize you might still be changing this, but in the event that you're not, I backported it to 0.11 along with the soft fork for BIP 68,112,113 here: 
https://github.com/bitcoin/bitcoin/compare/0.11...morcos:backportBIP9SoftFork?expand=1

The RPC tests are annoying to backport, so I left 2 of them out, but tested using the recent RPC test bip68-112-113.py against the 0.11 binary.

sipa commented:
@morcos I'm fine with the current state.

jonasschnelli commented:
utACK

petertodd commented on src/consensus/params.h:
```diff
@@ -7,8 +7,29 @@
 
 namespace Consensus {
+
+enum DeploymentPos
```
Why is this called "DeploymentPos"? It's not a bit position; the actual bit positions have another layer of indirection in the BIP9Deployment structure: https://github.com/sipa/bitcoin/blob/8c74cedef53ab791ed333f25794f8b9d2e9f51aa/src/consensus/params.h#L26

We should change this to DeploymentId

petertodd replied:
Oh, actually, this means the warning code is broken...

laanwj replied:
'position' is a very ambigious, yes. It's almost never a good choice.
Made a similar comment here on another pull: https://github.com/bitcoin/bitcoin/pull/7541#discussion_r53135937
Agree 'Id' seems be a better fit here.

petertodd replied:
Alright, I'll do a pull-req to fix that - but next week after the long weekend!

jtimon replied:
It is the position in Consensus::Params::vDeployments. Id could have been another option, but O disagree there's anything to fix here. Seems too late for bike-shedding this.

MarcoFalke replied:
Adding a comment for doxygen would also help if the purpose is not immediately clear from the code alone.

laanwj replied:
I had the same confusion as @petertodd when initially reading the code, so for new developers it must be worse. I do think at least a comment should be added.

jtimon replied:
ACK on adding documentation.
